### PR TITLE
add IMQSteinKernel

### PR DESCRIPTION
**Pull request by martinjankowiak**

### CODE REVIEW

**Comment by martinjankowiak** [comment]
@ jacobrgardner @gpleiss here's another kernel to play around with

**Comment by martinjankowiak** [comment]
if anyone knows how to fix the doc build errors lmk, cuz i'm stymied

File: pyro/infer/svgd.py [hunk 1]
```diff
@@ -101,6 +103,81 @@ def log_kernel_and_grad(self, particles):
...
+        :param float beta: Kernel hyperparameter, defaults to -0.5.
+        :param float bandwidth_factor: Optional factor by which to scale the bandwidth
+        """
+        assert alpha > 0.0, "alpha must be positive."
+        assert beta < 0.0, "beta must be negative."
+        if bandwidth_factor is not None:
+            assert bandwidth_factor > 0.0, "bandwidth_factor must be positive."
+        self.alpha = alpha
+        self.beta = beta
+        self._bandwidth_factor = bandwidth_factor
```
Comment (fritzo) [inline pyro/infer/svgd.py]: This should be
```py
self.bandwidth_factor = bandwidth_factor
```

File: pyro/infer/svgd.py [hunk 2]
```diff
@@ -101,6 +103,81 @@ def log_kernel_and_grad(self, particles):
...
+    """
+    def __init__(self, alpha=0.5, beta=-0.5, bandwidth_factor=None):
+        """
+        :param float alpha: Kernel hyperparameter, defaults to 0.5.
+        :param float beta: Kernel hyperparameter, defaults to -0.5.
+        :param float bandwidth_factor: Optional factor by which to scale the bandwidth
+        """
+        assert alpha > 0.0, "alpha must be positive."
+        assert beta < 0.0, "beta must be negative."
+        if bandwidth_factor is not None:
```
Comment (fritzo) [inline pyro/infer/svgd.py]: I would move this check into the setter, as you've done in RBFKernel.

File: pyro/infer/svgd.py [hunk 3]
```diff
@@ -71,6 +71,8 @@ def __init__(self, bandwidth_factor=None):
         """
         :param float bandwidth_factor: Optional factor by which to scale the bandwidth
         """
+        if bandwidth_factor is not None:
+            assert bandwidth_factor > 0.0, "bandwidth_factor must be positive."
         self._bandwidth_factor = bandwidth_factor
```
Comment (fritzo) [inline pyro/infer/svgd.py]: Use the setter.

File: pyro/infer/svgd.py [hunk 4]
```diff
@@ -71,6 +71,8 @@ def __init__(self, bandwidth_factor=None):
         """
         :param float bandwidth_factor: Optional factor by which to scale the bandwidth
         """
+        if bandwidth_factor is not None:
```
Comment (fritzo) [inline pyro/infer/svgd.py]: Remove the check after switching to the setter below.

File: pyro/infer/svgd.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/infer/svgd.py]: ah thanks this setter stuff always confuses me

File: pyro/infer/svgd.py (hunk 4 above)
Comment (martinjankowiak) [reply pyro/infer/svgd.py]: done

**Comment by martinjankowiak** [comment]
ah, no, nothing in particular all the meat is in the new `log_kernel_and_grad`

**Comment by martinjankowiak** [comment]
@fritzo thanks for fixing the docs!

**Pull request merged**